src/gflownet/algo/graph_sampling.py

```diff
@@ -0,0 +1,140 @@
+            # Forward pass to get GraphActionCategorical
+            fwd_cat, log_reward_preds = model(self.ctx.collate(torch_graphs).to(dev), cond_info[not_done_mask])
+            if self.random_action_prob > 0:
+                masks = [1] * len(fwd_cat.logits) if fwd_cat.masks is None else fwd_cat.masks
+                # Device which graphs in the minibatch will get their action randomized
+                is_random_action = torch.tensor(
+                    self.rng.uniform(size=len(torch_graphs)) < self.random_action_prob, device=dev).float()
+                # Set the logits to some large value if they're not masked, this way the masked
+                # actions have no probability of getting sampled, and there is a uniform
+                # distribution over the rest
+                fwd_cat.logits = [
+                    # We don't multiply m by i on the right because we're assume the model forward()
+                    # method already does that
+                    is_random_action[b][:, None] * torch.ones_like(i) * m * 100 + i * (1 - is_random_action[b][:, None])
+                    for i, m, b in zip(fwd_cat.logits, masks, fwd_cat.batch)
+                ]
+            if self.sample_temp != 1:
+                sample_cat = copy.copy(fwd_cat)
+                sample_cat.logits = [i / self.sample_temp for i in fwd_cat.logits]
+                actions = sample_cat.sample()
+            else:
+                actions = fwd_cat.sample()
+            graph_actions = [self.ctx.aidx_to_GraphAction(g, a) for g, a in zip(torch_graphs, actions)]
+            log_probs = fwd_cat.log_prob(actions)
+            # Step each trajectory, and accumulate statistics
+            for i, j in zip(not_done(range(n)), range(n)):
+                fwd_logprob[i].append(log_probs[j].unsqueeze(0))
+                data[i]['traj'].append((graphs[i], graph_actions[j]))
+                # Check if we're done
+                if graph_actions[j].action is GraphActionType.Stop or t == self.max_len - 1:
+                    done[i] = True
+                    if self.sanitize_samples and not self.ctx.is_sane(graphs[i]):
+                        # check if the graph is sane (e.g. RDKit can
+                        # construct a molecule from it) otherwise
+                        # treat the done action as illegal
+                        data[i]['is_valid'] = False
+                else:  # If not done, try to step the self.environment
+                    gp = graphs[i]
+                    try:
+                        # self.env.step can raise AssertionError if the action is illegal
+                        gp = self.env.step(graphs[i], graph_actions[j])
+                        assert len(gp.nodes) <= self.max_nodes
+                    except AssertionError:
+                        done[i] = True
+                        data[i]['is_valid'] = False
+                        continue
```

Comment:
Why do you use a try except block here? It looks like all you want is to check that `len(gp.nodes) <= self.max_nodes` and if it is not, to indicate that the graph is invalid. This can be accomplished with a simple `if` statement.

Reply:
`self.env.step` itself is designed to raise AssertionError for a bad action. It would probably be more clear if we raised something like a `GraphActionError`. Noted for future PRs.